Animated sprites

seungahpyun opened the pull request:
## Changes

### 🎬 **Animated Sprite System**

#### **Core Animation Features**
* **Time-Based Animation**: Implemented frame cycling with configurable animation speed (0.15s intervals) using `mlx_get_time()`
* **Multi-Frame Support**: Added support for 6-frame tree animations with `TREE_FRAMES` constant and frame management
* **Animation State Management**: Enhanced `t_sprite` structure with `current_frame`, `last_frame_time`, and `frames[]` array for complete animation control

#### **Sprite Management System**
* **Map Integration**: Parse 'T' characters from map files and convert to sprite objects
* **Collision Detection**: Sprite-player collision system for realistic interaction
* **Minimap Visualization**: sprite position display on minimap with distinct coloring

### 🏗️ **Code Structure Refactoring**

#### **Modular Directory Organization**
* **Categorized Subdirectories**: Reorganized code into `common/`, `control/`, `parser/`, and `render/` with logical subdivisions (`core/`, `memory/`, `utils/`)
* **Sprite Module**: Created dedicated `src/render/sprite/` directory with specialized files:
  - `sprite_animation.c` - Animation update logic
  - `sprite_drawing.c` - Pixel-level rendering
  - `sprite_loader.c` - Texture management
  - `sprite_renderer.c` - Main rendering pipeline
  - `sprite_sorting.c` - Distance-based sorting

seungahpyun committed:
change texture

seungahpyun committed:
adding tree texture

seungahpyun committed:
Merge branch 'main' into animated-sprites

seungahpyun committed:
parsing sprite

seungahpyun committed:
seems working but shouldnt make window

seungahpyun committed:
sprite minimap rendered

seungahpyun committed:
sprite rendered

seungahpyun committed:
change calculation

seungahpyun committed:
texture scale

seungahpyun committed:
change texture

seungahpyun committed:
cant work into tree

seungahpyun committed:
basic norminette

seungahpyun committed:
norm pass

seungahpyun committed:
clean up

seungahpyun committed:
change rendering orger

seungahpyun committed:
safer map freeing

seungahpyun committed:
duplicated calculation remove

seungahpyun committed:
duplicated function delete

seungahpyun committed:
extract math part

seungahpyun committed:
make subdirectory and clean up

seungahpyun committed:
norminette

seungahpyun committed:
render_3d_projection delete

seungahpyun commented on include/render.h:
```diff
@@ -101,42 +104,100 @@ typedef struct s_render_data
+/*                                    CORE                                    */
+/* ========================================================================== */
+/* 3d_projection.c */
+void		render_3d_projection(t_game *game);
```
good catch. I update this :)

seungahpyun commented on Makefile:
```diff
@@ -57,17 +74,30 @@ SRC_FILES = $(SRC_DIR)/main.c \
+			$(RENDER_SPRITE_DIR)/sprite_animation.c \
+			$(RENDER_SPRITE_DIR)/sprite_drawing.c \
+			$(RENDER_SPRITE_DIR)/sprite_loader.c \
+			$(RENDER_SPRITE_DIR)/sprite_position.c \
```
this it not true. it is exist file!

lanfeitiao commented:
Hey @seungahpyun Quick observation: I noticed that sprite textures are hard-coded while other textures are all user-defined in the .cub file. Should we make sprite textures configurable for consistency, or is the current approach intentional for this implementation?

seungahpyun committed:
merge with main

seungahpyun committed:
norm

seungahpyun committed:
not working

seungahpyun committed:
working

seungahpyun committed:
fix basic norminette

seungahpyun committed:
norm pass

seungahpyun committed:
map change

seungahpyun committed:
delete dup func

seungahpyun committed:
delete space

seungahpyun commented:
Hey @lanfeitiao  now implements dynamic parsing of animated sprites from the map configuration file, replacing the hardcoded tree texture approach. The implementation adds support for configurable animated sprites with flexible frame counts.

### Key Changes

#### 1. Map Character Update
- **Changed**: Map character from `T` (tree) to `A` (animated sprite)
- **Benefit**: More generic and descriptive naming convention

#### 2. Configuration Format
- **New Format**: 
  ```
  ANIMATED ./textures/tree00.png
  ANIMATED ./textures/tree01.png
  ANIMATED ./textures/tree02.png
  ...
  ```
- **Flexibility**: Users can specify any number of frames within the allowed range

#### 3. Frame Count Validation
- **Minimum**: 2 frames (ensures animation)
- **Maximum**: 10 frames (prevents excessive memory usage)
- **Validation**: Proper error handling for invalid frame counts

#### 4. Parser Implementation
- Added `animated_sprite_parser.c` for handling ANIMATED directives
- Integrated sprite parsing into the main element parser
- Added validation for texture paths and frame consistency

### Technical Implementation

#### New Files Added:
- `src/parser/element/animated_sprite_parser.c`
- `src/parser/utils/sprite_utils.c`

#### Key Functions:
- `parse_animated_sprite()`: Parses individual ANIMATED lines
- `validate_animated_sprite_config()`: Validates the complete configuration
- `is_valid_frame_count()`: Ensures frame count is within bounds

#### Configuration Structure:
```c
typedef struct s_animated_sprite_config
{
    char    *paths[MAX_SPRITE_FRAMES];  // Frame texture paths
    int     frame_count;                // Total number of frames
} t_animated_sprite_config;
```

you were right about moving away from hardcoding! My original implementation was just hardcoded values, but this approach is so much cleaner and more maintainable. Thanks again! 👍

seungahpyun committed:
norm fix

seungahpyun committed:
seg fault fix

seungahpyun committed:
norm and space fix

seungahpyun committed:
adding big map

lanfeitiao reviewed (commented):
Hey @seungahpyun — amazing work! Super impressed that you got the sprite config and smooth rendering done so quickly 👏

Just a small suggestion: I noticed we have some duplicate functions for map coordinate validation — like `is_valid_map_position` (from collision_check), `is_valid_point` (from render_utils), and `is_valid_map_coord` (from common), which all seem to do the same thing. Same for `is_within_boundary` (in common) and `is_coordinate_valid` (in render_utils).

Would it make sense to consolidate them into one shared utility?

seungahpyun committed:
delete is_valid_map_position

seungahpyun committed:
delete dup function

lanfeitiao reviewed (approved):
Great! Merged!
